changesummary.py

```diff
@@ -0,0 +1,268 @@
+            with the same. The file `'allapis.summary'` is saved to the current
+            working directory.
+            args:
+                dataframe: a pandas dataframe containing summary change
+                    information for all discovery artifacts
+                directory: path where the summary file should be saved
+        """
+        dataframe['IsFeature'] = np.where( \
+            (dataframe['ChangeType'] == ChangeType.DELETED) | \
+                (dataframe['ChangeType'] == ChangeType.ADDED), True, np.nan)
+
+        dataframe['IsBreaking'] = np.where( \
+            (dataframe['ChangeType'] == ChangeType.DELETED), True, np.nan)
+
+        dataframe['IsFeatureAggregate'] = dataframe.groupby( \
+            'Name').IsFeature.transform(lambda x : x.any())
+
+        dataframe['IsBreakingAggregate'] = dataframe.groupby(\
+            'Name').IsBreaking.transform(lambda x : x.any())
+
+        dataframe['Summary'] = np.vectorize(self._build_summary_message)\
+                                            (dataframe['Name'],
+                                            dataframe['IsFeatureAggregate'],
+                                            dataframe['IsBreakingAggregate'])
+
+        with open("".join([directory,"/", "allapis.summary"]), 'w') as f:
+            f.writelines([summary_msg for summary_msg in dataframe.Summary.unique()])
+
+        return dataframe
+
+    def _write_verbose_changes_to_disk(self, dataframe, directory, summary_df):
+        """"Writes verbose information to file about changes made to discovery
+            artifacts based on the provided dataframe. A separate file is saved
+            for each api in the current working directory. The extension of the
+            files will be `'.verbose'`.
+
+            args:
+                dataframe: a pandas dataframe containing verbose change
+                    information for all discovery artifacts
+                directory: path where the summary file should be saved
+                summary_df: A dataframe containing a summary of the changes
+        """
+        verbose_changes = []
+        change_type_groups = dataframe[['Name','Version','ChangeType','Key']].groupby(['Name','Version','ChangeType'])
+
+        lastApi = ''
+        lastVersion = ''
+        lastType = ChangeType.UNKNOWN
+
+        f = None
+        for name, group in change_type_groups:
+            currentApi = name[0]
+            currentVersion = name[1]
+            currentType = name[2]
+
+            if lastApi != currentApi or lastVersion != currentVersion:
+                if f is not None:
+                    f.writelines(verbose_changes)
+                    f.close()
+
+                verbose_changes = []
+                filename = ".".join([currentApi, "verbose"])
+                f = open("".join([directory,"/", filename]), "w")
+                verbose_changes.append(summary_df[summary_df["Name"]==name[0]]["Summary"].iloc[0])
```

Comment:
I'm having trouble following what this line because I don't know pandas. Could you add a comment explaining what text is being added here?

Reply:
Done.